src/bindings/concept/type_.rs

```diff
@@ -0,0 +1,685 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+use std::ffi::c_char;
+
+use typeql_lang::pattern::Annotation;
+
+use super::{
+    concept::{
+        borrow_as_attribute_type, borrow_as_attribute_type_mut, borrow_as_entity_type, borrow_as_entity_type_mut,
+        borrow_as_relation_type, borrow_as_relation_type_mut, borrow_as_role_type, borrow_as_thing_type,
+        borrow_as_thing_type_mut,
+    },
+    ConceptIterator,
+};
+use crate::{
+    bindings::common::{
+        borrow, borrow_optional, ok_record_flatten, release_optional, release_string, string_array_view, string_view,
+        unwrap_or_default, unwrap_or_null, unwrap_to_c_string, unwrap_void,
+    },
+    concept::{Concept, Transitivity, Value, ValueType},
+    transaction::concept::api::{AttributeTypeAPI, EntityTypeAPI, RelationTypeAPI, RoleTypeAPI},
+    Transaction,
+};
+
+#[no_mangle]
+pub extern "C" fn thing_type_is_abstract(thing_type: *const Concept) -> bool {
+    borrow_as_thing_type(thing_type).is_abstract()
+}
+
+#[no_mangle]
+pub extern "C" fn thing_type_get_label(thing_type: *const Concept) -> *mut c_char {
+    release_string(borrow_as_thing_type(thing_type).label().to_owned())
+}
+
+#[no_mangle]
+pub extern "C" fn thing_type_is_deleted(transaction: *const Transaction<'static>, thing_type: *const Concept) -> bool {
+    unwrap_or_default(borrow_as_thing_type(thing_type).is_deleted(borrow(transaction)))
+}
+
+#[no_mangle]
+pub extern "C" fn thing_type_delete(transaction: *const Transaction<'static>, thing_type: *mut Concept) {
+    unwrap_void(borrow_as_thing_type_mut(thing_type).delete(borrow(transaction)))
+}
+
+#[no_mangle]
+pub extern "C" fn thing_type_set_label(
+    transaction: *const Transaction<'static>,
+    thing_type: *mut Concept,
+    new_label: *const c_char,
+) {
+    unwrap_void(borrow_as_thing_type_mut(thing_type).set_label(borrow(transaction), string_view(new_label).to_owned()))
+}
+
+#[no_mangle]
+pub extern "C" fn thing_type_set_abstract(transaction: *const Transaction<'static>, thing_type: *mut Concept) {
+    unwrap_void(borrow_as_thing_type_mut(thing_type).set_abstract(borrow(transaction)))
+}
+
+#[no_mangle]
+pub extern "C" fn thing_type_unset_abstract(transaction: *const Transaction<'static>, thing_type: *mut Concept) {
+    unwrap_void(borrow_as_thing_type_mut(thing_type).unset_abstract(borrow(transaction)))
+}
+
+#[no_mangle]
+pub extern "C" fn thing_type_get_owns(
+    transaction: *const Transaction<'static>,
+    thing_type: *const Concept,
+    value_type: *const ValueType,
+    transitivity: Transitivity,
+    annotations: *const *const c_char,
+) -> *mut ConceptIterator {
+    let annotations = string_array_view(annotations)
+        .map(|anno| match anno {
+            // FIXME TypeQL?
```

Comment:
what's this FIXME?

Reply:
Annotations need to be exposed over FFI as an enum. Best way would be by making the `typeql_lang::Annotation` `#[repr(C)]`. Currently this is passed by converting the Java annotation to `String`.